consensus/view_change.go

```diff
@@ -214,228 +200,47 @@ func (consensus *Consensus) onViewChange(msg *msg_pb.Message) {
-		consensus.getLogger().Debug().
-			Str("validatorPubKey", senderKey.Bytes.Hex()).
-			Msg("[onViewChange] Already Received M3(ViewID) message from the validator")
+	members := consensus.Decider.Participants()
+	consensus.vc.AddViewIDKeyIfNotExist(recvMsg.ViewID, members)
+
+	// do it once only per viewID/Leader
+	if err := consensus.vc.InitPayload(consensus.FBFTLog,
+		recvMsg.ViewID,
+		recvMsg.BlockNum,
+		consensus.priKey); err != nil {
+		consensus.getLogger().Error().Err(err).Msg("Init Payload Error")
 		return
 	}
-	viewIDHash := make([]byte, 8)
-	binary.LittleEndian.PutUint64(viewIDHash, recvMsg.ViewID)
-	if !recvMsg.ViewidSig.VerifyHash(recvMsg.SenderPubkey.Object, viewIDHash) {
-		consensus.getLogger().Warn().
-			Uint64("MsgViewID", recvMsg.ViewID).
-			Msg("[onViewChange] Failed to Verify M3 Message Signature")
+
+	err = consensus.vc.ProcessViewChangeMsg(consensus.FBFTLog, consensus.Decider, recvMsg)
+	if err != nil {
+		consensus.getLogger().Error().Err(err).
+			Uint64("viewID", recvMsg.ViewID).
+			Uint64("blockNum", recvMsg.BlockNum).
+			Str("msgSender", recvMsg.SenderPubkey.Bytes.Hex()).
+			Msg("Verify View Change Message Error")
 		return
 	}
-	consensus.getLogger().Info().
-		Str("validatorPubKey", senderKey.Bytes.Hex()).
-		Msg("[onViewChange] Add M3 (ViewID) type message")
-
-	consensus.viewIDSigs[recvMsg.ViewID][senderKey.Bytes.Hex()] = recvMsg.ViewidSig
-	// Set the bitmap indicating that this validator signed.
-	consensus.viewIDBitmap[recvMsg.ViewID].SetKey(recvMsg.SenderPubkey.Bytes, true)
-	consensus.getLogger().Info().
-		Int("have", len(consensus.viewIDSigs[recvMsg.ViewID])).
-		Int64("total", consensus.Decider.ParticipantsCount()).
-		Msg("[onViewChange]")
 
 	// received enough view change messages, change state to normal consensus
-	if consensus.Decider.IsQuorumAchievedByMask(consensus.viewIDBitmap[recvMsg.ViewID]) {
+	if consensus.Decider.IsQuorumAchievedByMask(consensus.vc.GetViewIDBitmap(recvMsg.ViewID)) {
+		consensus.getLogger().Info().Msg("[onViewChange] View Change Message Quorum Reached")
 		consensus.current.SetMode(Normal)
 		consensus.LeaderPubKey = newLeaderKey
 		consensus.ResetState()
-		if len(consensus.m1Payload) == 0 {
+		if consensus.vc.IsM1PayloadEmpty() {
 			// TODO(Chao): explain why ReadySignal is sent only in this case but not the other case.
 			// Make sure the newly proposed block have the correct view ID
 			consensus.SetCurBlockViewID(recvMsg.ViewID)
 			go func() {
 				consensus.ReadySignal <- struct{}{}
 			}()
 		} else {
-			consensus.getLogger().Info().
-				Str("From", consensus.phase.String()).
-				Str("To", FBFTCommit.String()).
-				Msg("[OnViewChange] Switching phase")
-			consensus.switchPhase(FBFTCommit, true)
-			copy(consensus.blockHash[:], consensus.m1Payload[:32])
-			aggSig, mask, err := consensus.ReadSignatureBitmapPayload(consensus.m1Payload, 32)
+			consensus.switchPhase("onViewChange", FBFTCommit)
+			payload := consensus.vc.GetM1Payload()
```

Comment:
this has potential race condition.

Reply:
we can lock the entire quorum reach block though.  what's the potential race condition here?